### ISSUE TITLE
Alerting: Alerts invisible after legacy migration after previous opt-out

### ISSUE THREAD

**Issue by PhilippWendler** [grafana/grafana#71529]
### What happened?

When we upgraded to Grafana 9.x, our existing legacy alerts were migrated to the new alerting. AFAIR this was successful and I could see the alerts in the Grafana UI. However, we soon [rolled back](https://grafana.com/docs/grafana/v9.5/alerting/migrating-alerts/roll-back/) as described in the documentation and continued to use the legacy alerting. This also worked fine until now. Now we are ready to migrate to the unified alerting. So I configured Grafana as described for [opt-in](https://grafana.com/docs/grafana/v9.5/alerting/migrating-alerts/opt-in/) and started it again.

The problem is that I now cannot see any alert rules in the Grafana UI. The page `/alerting/list` just shows an empty list. Those panels that had an alert configured are now shown as having no alert.

Contact points and notification policies are shown.

Interestingly, despite no alert rule at all shown in UI, at least some of the rules are still active, because I do get notifications. The link behind "View alert" in the notification leads me to a page that shows "Rule not found":
![image](https://github.com/grafana/grafana/assets/2545335/dd864301-83b0-407f-9ef2-f14ec762a767)

On startup, Grafana logs the following messages (both on the first start after the opt-in and on further startups):
```
logger=settings t=2023-07-13T10:28:05.348074374+02:00 level=warn msg="\"sentry\" frontend logging provider is deprecated and will be removed in the next major version. Use \"grafana\" provider instead."
logger=ngalert t=2023-07-13T10:28:06.586750451+02:00 level=warn msg="Unexpected number of rows updating alert configuration history" rows=0 org=1 hash=not-yet-calculated
logger=modules t=2023-07-13T10:28:07.811053743+02:00 level=warn msg="No modules registered..."
```

It seems to me that maybe there are some left-overs from the first migration in the database or so and that because of this the current migration is not performed? I expected to see log messages regarding the migration after the opt-in, but there were none. Can I somehow clean up the database such that the migration is performed (again)?

### What did you expect to happen?

I expect all my alert rules to be migrated and shown in the UI.

### Did this work before?

This is the first time I try the migration after the previous rollback.

### How do we reproduce it?

1. Upgrade triggering migration
2. Rollback
3. Opt-in to migration

Right now I can repeat the attempted opt-in migration as often as I wish as I have a backup from the Grafana state before the opt-in. So if there is anything I can provide to assist with this, I would be glad to do so. I do not have anything from the time of the first migration and the rollback, though.

### Environment (with versions)?

Grafana: 9.5.6
OS: Ubuntu 22.04
Browser: Firefox 115.0.2

### Grafana platform?

A package manager (APT, YUM, BREW, etc.)

**Comment by JacobsonMT** [comment]
Hi @PhilippWendler, thanks for the bug report! Some things that could help us track this down are:

1. What does the export of all alert rules look like? You can access this via `https://<grafana>/api/v1/provisioning/alert-rules/export?download=false&format=yaml` and make sure to redact if there's anything confidential.
2. In the past we've seen odd legacy database states affect the migration in similar ways. For example, when legacy alerts had missing dashboards, or the dashboard was orphaned from its folder, or the entire org no longer existed. One thing to look at would be to confirm that the legacy alert rules seem to be well formed in this respect.
3. Could you provide the migration logs from any of the attempts? They should look something like this:

```
INFO [07-14|12:18:00] Connecting to DB                         logger=sqlstore dbtype=mysql
INFO [07-14|12:18:00] Starting DB migrations                   logger=migrator
INFO [07-14|12:18:00] Executing migration                      logger=migrator id="drop non-unique orgID index on alert_configuration"
INFO [07-14|12:18:00] Executing migration                      logger=migrator id="drop unique orgID index on alert_configuration if exists"
WARN [07-14|12:18:00] Skipping migration: Already executed, but not recorded in migration log logger=migrator id="drop unique orgID index on alert_configuration if exists"
INFO [07-14|12:18:00] Executing migration                      logger=migrator id="extract alertmanager configuration history to separate table"
INFO [07-14|12:18:00] Executing migration                      logger=migrator id="add unique index on orgID to alert_configuration"
INFO [07-14|12:18:00] Executing migration                      logger=migrator id="clear migration entry \"remove unified alerting data\""
INFO [07-14|12:18:00] Executing migration                      logger=migrator id="move dashboard alerts to unified alerting"
INFO [07-14|12:18:00] alerts found to migrate                  logger=migrator alerts=15
INFO [07-14|12:18:00] Executing migration                      logger=migrator id="clone move dashboard alerts to unified alerting"
INFO [07-14|12:18:00] Executing migration                      logger=migrator id="update dashboard_uid and panel_id from existing annotations"
INFO [07-14|12:18:00] Executing migration                      logger=migrator id="create default alerting folders"
INFO [07-14|12:18:00] Executing migration                      logger=migrator id="update group index for alert rules"
```

To answer your specific questions:

> It seems to me that maybe there are some left-overs from the first migration in the database or so and that because of this the current migration is not performed? 

This is possible but unlikely since the migration roll back will [clear all tables](https://github.com/grafana/grafana/blob/3af95bebe1600d3c27668359998a2538d85e1982/pkg/services/sqlstore/migrations/ualert/ualert.go#L540-L599) related to the new alerting configuration. 

>Can I somehow clean up the database such that the migration is performed (again)?

I notice you mention that you have a database backup from pre-migration, are you restoring the back to re-run the migration or are you rolling back. If you're restoring, it might be worth trying to rollback to let it clear the tables.

> I expected to see log messages regarding the migration after the opt-in, but there were none.

Were there logs from other migrations (not related to upgrading alerting)?

**Comment by PhilippWendler** [comment]
Thanks for responding!

>   1. What does the export of all alert rules look like? You can access this via `https://<grafana>/api/v1/provisioning/alert-rules/export?download=false&format=yaml` and make sure to redact if there's anything confidential.

Wow, thanks to this I think I found it out! This URL gave me a "missing permission" error. So I played around with the Grafana config in order to see what could give me the required permission, and finally set `org_role = Admin` in `[auth.anonymous]` (I have no users and no authentication configured and only use anonymous access with `org_role = Editor` usually.) That allowed me to access the export, but luckily I noticed that in the other browser tab now suddenly my alert rules where visible!

So `org_role = Editor` gives me an empty list of rules, but `org_role = Admin` allows me to see them and this is reproducible. Turns out the migration was working fine all the time. :-)

However, I am surprised that I need to be admin to see the rules, is this intended? It always worked fine as editor for the legacy alerting, and even with unified alerting I can still see contact points etc. as editor, just not the rules.

Or can I give editor somehow the permissions to see (and edit!) the rules?

---

For completeness:

>  3. Could you provide the migration logs from any of the attempts? They should look something like this:

Ah, I am sorry. I had set `level = debug` in `[log]` but had overlooked that I had `level = warn` in `[log.syslog]`, that's why the log above has only warnings. I have now produced a debug log for a fresh start of Grafana with unified alerting enabled after restoring my backed-up state (i.e., the one with rolled-back alerts), so Grafana should run a migration here: 
[grafana.log](https://github.com/grafana/grafana/files/12052638/grafana.log)

It contains this:
```
logger=migrator t=2023-07-14T19:30:01.754298776+02:00 level=info msg="Executing migration" id="clear migration entry \"remove unified alerting data\""
logger=migrator t=2023-07-14T19:30:01.754305936+02:00 level=debug msg="Executing code migration" id="clear migration entry \"remove unified alerting data\""
logger=migrator t=2023-07-14T19:30:01.754384723+02:00 level=info msg="Executing migration" id="move dashboard alerts to unified alerting"
logger=migrator t=2023-07-14T19:30:01.754393233+02:00 level=debug msg="Executing code migration" id="move dashboard alerts to unified alerting"
logger=migrator t=2023-07-14T19:30:01.756830069+02:00 level=info msg="alerts found to migrate" alerts=25
```

**Comment by PhilippWendler** [comment]
Is there a way to make non-admins be able to configure alert rules? Seems like a regression compared to the legacy alerting, where this was possible.

**Comment by JacobsonMT** [comment]
> So org_role = Editor gives me an empty list of rules, but org_role = Admin allows me to see them and this is reproducible. Turns out the migration was working fine all the time. :-)

I'm glad the migration was working as intended, though there does seem to be something else weird going on here. Editor roles should be able to see and edit alert rules by default. I think there might be a bug in the migration to do with permissions on the folder used to migrate legacy alerts that were in the `General` folder.

Were the legacy alerts you migrated in the `General` folder / are they now in the `General Alerting` folder?

**Comment by PhilippWendler** [comment]
> Were the legacy alerts you migrated in the `General` folder / are they now in the `General Alerting` folder?

Yes.

> If so, a quick fix for you will be to go to the `General Alerting` folder permissions tab and add permissions for Role Viewer and Role Editor, something like:

Ok, it took me some time to find that page. :-) Is it really only accessible via Home -> Dashboards -> Hover over folder -> Go to folder? I kept looking for something about folders or permissions in the menu, but didn't find anything.

But that's just a side note, I now added edit permissions for Editor and it seems to work. I even created a new rule and got an "Unauthorized" notification bubble after saving it, but it was still created.

Thanks a lot for your help!
